doc: IsFinalTx comment about nSequence & OP_CLTV

nothingmuch opened the pull request:
It's somewhat surprising that a transaction's `nLockTime` field is ignored
when all `nSequence` fields are final, so this change aims to clarify this
behavior and cross reference relevant details of `OP_CHECKLOCKTIMEVERIFY`.

nothingmuch committed:
doc: IsFinalTx comment about nSequence & OP_CLTV

It's somewhat surprising that a transaction's nLockTime field is ignored
when all nSequence fields are final, so this change aims to clarify this
behavior and cross reference relevant details of OP_CHECKLOCKTIMEVERIFY.

amitiuttarwar commented:
ACK 5546ed61c65c167ba839df7b5277f8689242ae8b. Sheds light on a confusing corner of the code.

ariard commented on src/script/interpreter.cpp:
```diff
@@ -1347,9 +1347,9 @@ bool GenericTransactionSignatureChecker<T>::CheckLockTime(const CScriptNum& nLoc
+    // and thus CHECKLOCKTIMEVERIFY bypassed if every txin has
+    // been finalized by setting nSequence to maxint. The
     // transaction would be allowed into the blockchain, making
     // the opcode ineffective.
```
"This enable a lightweight onchain covenant where you can be safe than constrained output can't be spent until reaching committed nBlockHeight/nBlockTime"

nothingmuch replied:
I expanded a bit on your suggestion in c4a022d4

p.s. sorry for accidentally clicking resolve

ariard reviewed (approved):
ACK 5546ed6

Thanks for documenting this foresight of OP_CLTV design.

michaelfolkson commented on src/consensus/tx_verify.cpp:
```diff
@@ -20,6 +20,15 @@ bool IsFinalTx(const CTransaction &tx, int nBlockHeight, int64_t nBlockTime)
         return true;
+
+    // Even if tx.nLockTime isn't satisfied by nBlockHeight/nBlockTime, a
+    // transaction is still considered final if all nSequence values are final,
```
nit: You could make clear that by final you mean if the nSequence values are maxed out to `0xffffffff`. Perhaps this wording of final is clear to everyone else (in which case ignore).

michaelfolkson reviewed (commented):
ACK c4a022d4f75384ba9d6978971fcaeabd072a9acc. Nice addition.

amitiuttarwar commented:
@nothingmuch could you squash the fixup commit ?

nothingmuch commented:
Done. I've also included @michaelfolkson's suggestion.

nothingmuch commented:
After rereading the diff I added a minor rewording of the comment explaining nLockTime covenants, since I felt it lacked clarity, and force pushed that so that the PR is still squashed at the moment.

michaelfolkson commented:
ACK ae93eb3379604a5ee83242f420bb5431fa238b9e. British English spelling of "behaviour" but not worth another force push to change in my view.

amitiuttarwar commented on src/script/interpreter.cpp:
```diff
@@ -1434,16 +1434,22 @@ bool GenericTransactionSignatureChecker<T>::CheckLockTime(const CScriptNum& nLoc
-    // Finally the nLockTime feature can be disabled and thus
-    // CHECKLOCKTIMEVERIFY bypassed if every txin has been
-    // finalized by setting nSequence to maxint. The
+    // Finally the nLockTime feature can be disabled in IsFinalTx()
```
I'm not a huge fan of having function names in the comments, because the code might change over time. eg. theoretically there could be another calling path that also disables, right? 

in this case I don't think its a big deal esp bc its consensus code that seems _quite unlikely_ to change

nothingmuch replied:
FWIW the rationale for adding it was specifically to cross reference these the two locations in the code, as they are more easily understood together.

I can definitely see why this is not worth the risk of stale comments, but couldn't think of a simpler way.

Maybe just mentioning `CHECKLOCKTIMEVERIFY` and `SEQUENCE_FINAL` is a good compromise, there are few false positives, and those are even less likely to change.

amitiuttarwar replied:
> FWIW the rationale for adding it was specifically to cross reference these the two locations in the code, as they are more easily understood together.

fair enough. makes sense in this case

amitiuttarwar reviewed (commented):
ACK ae93eb3379604a5ee83242f420bb5431fa238b9e

one comment, but not worth holding up the PR. its been open for a while and overall this documentation helps clarify some unexpected logic, so def an improvement in my eyes.

theStack commented:
Concept ACK, definitely makes sense to document unexpected behaviour in more detail.

nothingmuch commented:
i fixed both "behaviour" and "the the", will squash & rebase after i reread everything tomorrow

nothingmuch commented:
so, despite previous ACKs i couldn't resist rephrasing the paragraph about the locktime covenants, since i could barely understand what i had originally meant by it... squashed and rebased on master

amitiuttarwar commented:
ACK abc8f3bae2cd9a65e170528d77973307050c5b47.

Hoping this PR can get merged soon!

benthecarman commented:
ACK `abc8f3b`

Zero-1729 commented:
ACK abc8f3b. Looks good.

MarcoFalke commented on src/script/interpreter.cpp:
```diff
@@ -1434,16 +1434,23 @@ bool GenericTransactionSignatureChecker<T>::CheckLockTime(const CScriptNum& nLoc
+    // be unspendable until a specific nBlockHeight/nBlockTime
+    // committed in the scriptPubKey, as opposed to a spending
+    // transaction's nLockTime only being constrained by
+    // signature verification.
```
I think the Bitcoin Core source code (or interpreter source code) is the wrong place to document what is possible with each opcode. Better places to document this is the BIPs repo (or stackexchange, the mailing list, your private blog or website, ...).

In fact the BIP already mentions several use cases. See https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki#Motivation . Should the other use-cases be copy-pasted here as well? If a new use-case is found, should it be copied here, too? How much review/maintenance time is this going to consume?

Also, you mention the word `covenant`, which isn't defined in our source code, nor the BIP text. I think this section adds more confusion for anyone reading the BIP and then this section.

nothingmuch replied:
That sounds reasonable to me and fits in with the original motivation. @ariard since this was your [suggestion](https://github.com/bitcoin/bitcoin/pull/18096#discussion_r379129481) do you mind if I remove it?

nothingmuch replied:
I went ahead and removed this addition.

ariard replied:
Fine to remove it. IIRC, I did advise you this comment because I was surprised to see nowhere this relation between nLocktime enforcement and finality of inputs documented. At least not in BOLT3, where timeout transactions nSequence are carefully set to non-final.

Overall I agree with @MarcoFalke. Though I would say some mechanism usage towards specific Bitcoin applications are already documented in the codebase. For e.g the carve-out rule in the mempool : https://github.com/bitcoin/bitcoin/blob/683dda2a70e7a210996fa34be23bd0c563971ba9/src/validation.cpp#L801

maflcko reviewed (dismissed):
Seems fine to better explain how IsFinalTx works, but I am not sure about starting a section about "covenants".

78051301012 reviewed (commented):
src/script/interpreter.cpp

ariard commented on src/consensus/tx_verify.cpp:
```diff
@@ -20,6 +20,15 @@ bool IsFinalTx(const CTransaction &tx, int nBlockHeight, int64_t nBlockTime)
         return true;
+
+    // Even if tx.nLockTime isn't satisfied by nBlockHeight/nBlockTime, a
+    // transaction is still considered final if all nSequence == SEQUENCE_FINAL
```
nit: if all inputs's nSequence

ariard reviewed (commented):
ACK 047c607

MarcoFalke commented:
ACK f9e37f33ce2d8b463a0bcbe7189c9bc5b36530b7
